fdbserver/CommitProxyServer.actor.cpp

```diff
@@ -1179,6 +1179,70 @@ void determineCommittedTransactions(CommitBatchContext* self) {
 	}
 }
 
+inline int64_t extractTenantIdFromStringRef(StringRef s) {
+	if (s.size() < TenantAPI::PREFIX_SIZE) {
+		return TenantInfo::INVALID_TENANT;
+	}
+	// Parse mutation key to determine mutation encryption domain
+	StringRef prefix = s.substr(0, TenantAPI::PREFIX_SIZE);
+	return TenantAPI::prefixToId(prefix, EnforceValidTenantId::False);
+}
+
+int64_t extractTenantIdFromSingleKeyMutation(MutationRef m) {
+	ASSERT_WE_THINK(!isSystemKey(m.param1));
+	ASSERT_WE_THINK(isSingleKeyMutation((MutationRef::Type)m.type));
+
+	return extractTenantIdFromStringRef(m.param1);
+}
+
+// Check a single-key mutation is associated with a valid tenant id
+// FIXME(xwang): handle clear range properly
+ErrorOr<Void> checkTenantModeAccess(MutationRef m) {
```

Comment:
Nit: This could probably be written to return just a boolean since the only options are to return Void or `illegal_tenant_access`.